### PR TITLE
Add Dockerfiles for alpine3.18

### LINKED ISSUE #4607: Add Dockerfiles for Alpine 3.18
Now that [Alpine 3.18 has been released](https://alpinelinux.org/posts/Alpine-3.18.0-released.html), we should add Dockerfiles for it and get things tested.

### PR DESCRIPTION
**Pull request by lbussell**
Fixes #4607 

Release issue: https://github.com/dotnet/release/issues/541

### CODE REVIEW

File: README.aspnet.md [hunk 1]
```diff
@@ -72,19 +74,21 @@ Tags | Dockerfile | OS Version
...
 
 ## Linux arm64 Tags
 Tags | Dockerfile | OS Version
 -----------| -------------| -------------
 7.0.5-bullseye-slim-arm64v8, 7.0-bullseye-slim-arm64v8, 7.0.5-bullseye-slim, 7.0-bullseye-slim, 7.0.5, 7.0 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/aspnet/7.0/bullseye-slim/arm64v8/Dockerfile) | Debian 11
-7.0.5-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0-alpine-arm64v8, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/aspnet/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
+7.0.5-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/aspnet/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
+7.0.5-alpine3.18-arm64v8, 7.0-alpine3.18-arm64v8, 7.0-alpine-arm64v8, 7.0.5-alpine3.18, 7.0-alpine3.18 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/aspnet/7.0/alpine3.18/arm64v8/Dockerfile) | Alpine 3.18
```
Comment (MichaelSimons) [inline README.aspnet.md:86]: The 7.0-alpine-arm64v8 floating tag should stay pointing to alpine 3.17

File: README.runtime.md [hunk 2]
```diff
@@ -55,11 +55,13 @@ docker run --rm mcr.microsoft.com/dotnet/samples
 Tags | Dockerfile | OS Version
 -----------| -------------| -------------
 7.0.5-bullseye-slim-amd64, 7.0-bullseye-slim-amd64, 7.0.5-bullseye-slim, 7.0-bullseye-slim, 7.0.5, 7.0 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/bullseye-slim/amd64/Dockerfile) | Debian 11
-7.0.5-alpine3.17-amd64, 7.0-alpine3.17-amd64, 7.0-alpine-amd64, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/alpine3.17/amd64/Dockerfile) | Alpine 3.17
```
Comment (MichaelSimons) [inline README.runtime.md:58]: Looks like the 7.0-alpine-amd64 tag was dropped.

File: README.runtime.md [hunk 3]
```diff
@@ -68,19 +70,21 @@ Tags | Dockerfile | OS Version
...
 8.0.0-preview.5-jammy-chiseled-amd64, 8.0-preview-jammy-chiseled-amd64, 8.0.0-preview.5-jammy-chiseled, 8.0-preview-jammy-chiseled | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/8.0/jammy-chiseled/amd64/Dockerfile) | Ubuntu 22.04
 
 ## Linux arm64 Tags
 Tags | Dockerfile | OS Version
 -----------| -------------| -------------
 7.0.5-bullseye-slim-arm64v8, 7.0-bullseye-slim-arm64v8, 7.0.5-bullseye-slim, 7.0-bullseye-slim, 7.0.5, 7.0 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/bullseye-slim/arm64v8/Dockerfile) | Debian 11
-7.0.5-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0-alpine-arm64v8, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
+7.0.5-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
```
Comment (MichaelSimons) [inline README.runtime.md:81]: Looks like the 7.0-alpine-arm64v8 tag was dropped.

File: README.runtime.md [hunk 4]
```diff
@@ -89,19 +93,21 @@ Tags | Dockerfile | OS Version
...
 8.0.0-preview.5-jammy-chiseled-arm64v8, 8.0-preview-jammy-chiseled-arm64v8, 8.0.0-preview.5-jammy-chiseled, 8.0-preview-jammy-chiseled | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/8.0/jammy-chiseled/arm64v8/Dockerfile) | Ubuntu 22.04
 
 ## Linux arm32 Tags
 Tags | Dockerfile | OS Version
 -----------| -------------| -------------
 7.0.5-bullseye-slim-arm32v7, 7.0-bullseye-slim-arm32v7, 7.0.5-bullseye-slim, 7.0-bullseye-slim, 7.0.5, 7.0 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/bullseye-slim/arm32v7/Dockerfile) | Debian 11
-7.0.5-alpine3.17-arm32v7, 7.0-alpine3.17-arm32v7, 7.0-alpine-arm32v7, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/alpine3.17/arm32v7/Dockerfile) | Alpine 3.17
+7.0.5-alpine3.17-arm32v7, 7.0-alpine3.17-arm32v7, 7.0.5-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/7.0/alpine3.17/arm32v7/Dockerfile) | Alpine 3.17
```
Comment (MichaelSimons) [inline README.runtime.md:104]: Looks like the 7.0-alpine-arm32v7 tag was dropped.

File: README.sdk.md [hunk 5]
```diff
@@ -65,53 +65,59 @@ Tags | Dockerfile | OS Version
...
 8.0.100-preview.5-jammy-amd64, 8.0-preview-jammy-amd64, 8.0.100-preview.5-jammy, 8.0-preview-jammy | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/sdk/8.0/jammy/amd64/Dockerfile) | Ubuntu 22.04
 
 ## Linux arm64 Tags
 Tags | Dockerfile | OS Version
 -----------| -------------| -------------
 7.0.302-bullseye-slim-arm64v8, 7.0-bullseye-slim-arm64v8, 7.0.302-bullseye-slim, 7.0-bullseye-slim, 7.0.302, 7.0 | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/sdk/7.0/bullseye-slim/arm64v8/Dockerfile) | Debian 11
-7.0.302-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0-alpine-arm64v8, 7.0.302-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/sdk/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
+7.0.302-alpine3.17-arm64v8, 7.0-alpine3.17-arm64v8, 7.0.302-alpine3.17, 7.0-alpine3.17, 7.0-alpine | [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/nightly/src/sdk/7.0/alpine3.17/arm64v8/Dockerfile) | Alpine 3.17
```
Comment (MichaelSimons) [inline README.sdk.md:87]: Looks like the 7.0-alpine-arm64v8 tag was dropped.

File: tests/Microsoft.DotNet.Docker.Tests/TestData.cs [hunk 6]
```diff
@@ -35,10 +36,13 @@ public static class TestData
             new ProductImageData { Version = V6_0, OS = OS.JammyChiseled,       Arch = Arch.Arm64,   SdkOS = OS.Jammy },
             new ProductImageData { Version = V6_0, OS = OS.Alpine317,           Arch = Arch.Arm },
             new ProductImageData { Version = V6_0, OS = OS.Alpine317,           Arch = Arch.Arm64 },
+            new ProductImageData { Version = V6_0, OS = OS.Alpine318,           Arch = Arch.Arm },
+            new ProductImageData { Version = V6_0, OS = OS.Alpine318,           Arch = Arch.Arm64 },
             new ProductImageData { Version = V7_0, OS = OS.BullseyeSlim,        Arch = Arch.Amd64 },
             new ProductImageData { Version = V7_0, OS = OS.Jammy,               Arch = Arch.Amd64 },
             new ProductImageData { Version = V7_0, OS = OS.JammyChiseled,       Arch = Arch.Amd64,   SdkOS = OS.Jammy },
-            new ProductImageData { Version = V7_0, OS = OS.Alpine317,           Arch = Arch.Amd64 },
```
Comment (MichaelSimons) [inline tests/Microsoft.DotNet.Docker.Tests/TestData.cs:41]: How come this 7.0 entry was removed?

**Comment by mthalman** [comment]
Looks like there's a PowerShell failure in 6.0 that needs to be investigated. Seems consistent.

**Comment by lbussell** [comment]
@jander-msft, when do we want to move monitor to alpine3.18? Sometime after these are released and you have time to test it?

**Comment by jander-msft** [comment]
> @jander-msft, when do we want to move monitor to alpine3.18? Sometime after these are released and you have time to test it?

It should move to 3.18 when the floating alpine tags are moved to 3.18

**Comment by jander-msft** [comment]
> > @jander-msft, when do we want to move monitor to alpine3.18? Sometime after these are released and you have time to test it?
> 
> It should move to 3.18 when the floating alpine tags are moved to 3.18

Completely separate from this effort to add the 3.18, I would like to somehow codify that .NET Monitor should move with the floating tags in the manifest and dockerfiles such that we don't have to think about it, but I haven't worked that out yet. We did briefly discuss this earlier this year, and it might require a new set of floating tags. I can log an issue about this where it can ideated before executing on it.

File: eng/mcr-tags-metadata-templates/aspnet-tags.yml [hunk 7]
```diff
@@ -1,52 +1,58 @@
...
-$(McrTagsYmlTagGroup:8.0-preview-alpine3.17-amd64)
+$(McrTagsYmlTagGroup:8.0-preview-alpine3.18-amd64)
     customSubTableTitle: .NET 8 Preview Tags
 $(McrTagsYmlTagGroup:8.0-preview-jammy-amd64)
     customSubTableTitle: .NET 8 Preview Tags
 $(McrTagsYmlTagGroup:8.0-preview-jammy-chiseled-amd64)
     customSubTableTitle: .NET 8 Preview Tags
 $(McrTagsYmlTagGroup:7.0-bullseye-slim-amd64)
 $(McrTagsYmlTagGroup:7.0-alpine3.17-amd64)
+$(McrTagsYmlTagGroup:7.0-alpine3.18-amd64)
```
[the remark this answers is not in the document]
Comment (mthalman) [reply eng/mcr-tags-metadata-templates/aspnet-tags.yml:11]: This requires regenerating the readmes.

**Comment by lbussell** [comment]
The issue here looks similar to https://github.com/PowerShell/PowerShell/issues/19679 but with a different type initializer: `System.Management.Automation.TypeAccelerators` vs. `System.Management.Automation.PSVersionInfo`.

```bash
System.InvalidOperationException : Failed to execute docker run --name 6.0-pwsh-133292649288916909 --rm  mcr.microsoft.com/dotnet/nightly/sdk:6.0-alpine3.18-amd64 pwsh -c (Get-Childitem env:DOTNET_RUNNING_IN_CONTAINER).Value
Exit code: 139
Standard Error: One or more errors occurred. (The type initializer for 'System.Management.Automation.TypeAccelerators' threw an exception.)
Process terminated. The type initializer for 'System.Management.Automation.PSObject' threw an exception.
at System.Environment.FailFast(System.String, System.Exception)
at Microsoft.PowerShell.UnmanagedPSEntry.Start(System.String[], Int32)
at Microsoft.PowerShell.ManagedPSEntry.Main(System.String[])
System.TypeInitializationException: The type initializer for 'System.Management.Automation.PSObject' threw an exception.
---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.


File name: 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
--- End of inner exception stack trace ---
at System.Management.Automation.PSObject.get_Properties()
at System.Management.Automation.HostUtilities.GetDollarProfile(String allUsersAllHosts, String allUsersCurrentHost, String currentUserAllHosts, String currentUserCurrentHost)
at Microsoft.PowerShell.ConsoleHost.DoRunspaceInitialization(Boolean skipProfiles, String initialCommand, String configurationName, Collection1 initialCommandArgs)    at Microsoft.PowerShell.ConsoleHost.DoCreateRunspace(String initialCommand, Boolean skipProfiles, Boolean staMode, String configurationName, Collection1 initialCommandArgs)
at Microsoft.PowerShell.ConsoleHost.CreateRunspace(Object runspaceCreationArgs)
at Microsoft.PowerShell.ConsoleHost.DoRunspaceLoop(String initialCommand, Boolean skipProfiles, Collection`1 initialCommandArgs, Boolean staMode, String configurationName)
at Microsoft.PowerShell.ConsoleHost.Run(CommandLineParameterParser cpp, Boolean isPrestartWarned)
at Microsoft.PowerShell.ConsoleHost.Start(String bannerText, String helpText)
at Microsoft.PowerShell.UnmanagedPSEntry.Start(String[] args, Int32 argc)
```

**Comment by lbussell** [comment]
I think we'll need to wait on https://github.com/dotnet/runtime/pull/86284 to flow through. It might already be in the nightly builds but since https://github.com/dotnet/dotnet-docker/pull/4628 is blocked, I'm not sure yet.

There's also a full release checklist here: https://github.com/dotnet/core/issues/8470

**Comment by mthalman** [comment]
> It might already be in the nightly builds but since #4628 is blocked, I'm not sure yet.

Since this is 6.0, it wouldn't be there until June servicing.

**Comment by lbussell** [comment]
I filed https://github.com/PowerShell/PowerShell/issues/19703.

> Since this is 6.0, it wouldn't be there until the June servicing release.

What's the resolution then? If the RID is the solution (which it may or may not be) would we have to wait until July to ship Alpine3.18 for 6.0? Or just validate internally ahead of time?

**Comment by mthalman** [comment]
> What's the resolution then? If the RID is the solution (which it may or may not be) would we have to wait until July to ship Alpine3.18 for 6.0? Or just validate internally ahead of time?

Assuming the RID fixes the issue, then yes, we'd have to wait until July. We should be able to validate that in our internal branch now.

**Comment by lbussell** [comment]
An internal test run validated that the RID addition solves the PowerShell issue: https://dev.azure.com/dnceng/internal/_build/results?buildId=2187733&view=results (internal MSFT link)

**Comment by tvbishan** [comment]
@lbussell @mthalman So, we have to wait until July to receive Alpine 3.18 for 6.0?

**Comment by mthalman** [comment]
> @lbussell @mthalman So, we have to wait until July to receive Alpine 3.18 for 6.0?

That hasn't been determined yet.

File: tests/Microsoft.DotNet.Docker.Tests/SdkImageTests.cs [hunk 8]
```diff
@@ -32,11 +32,20 @@ public SdkImageTests(ITestOutputHelper outputHelper)
...
+                if (imageData.IsArm)
+                {
+                    // PowerShell needs support for Arm-based Alpine (https://github.com/PowerShell/PowerShell/issues/14667, https://github.com/PowerShell/PowerShell/issues/12937)
+                    reason = "PowerShell does not have Alpine arm images, skip testing";
+                    return false;
+                }
+                else if (imageData.OS.Contains("3.18"))
+                {
+                    // PowerShell does not support Alpine 3.18 yet (https://github.com/PowerShell/PowerShell/issues/19703)
+                    reason = "Powershell does not support Alpine 3.18 yet, skip testing"
```
Comment (mthalman) [inline tests/Microsoft.DotNet.Docker.Tests/SdkImageTests.cs:46]:
```suggestion
                    reason = "Powershell does not support Alpine 3.18 yet, skip testing";
```

**Review by mthalman — COMMENTED** [review]
Looks like the Dockerfiles need to be regenerated.

**Review by mthalman — APPROVED** [review]

**Review by MichaelSimons — APPROVED** [review]

**Pull request merged**